### ISSUE TITLE
Why are hard-coded setting IDs used?

### ISSUE THREAD

**Issue by chrismacklin** [labrad/pylabrad#145]
I understand the need for uniquely-numbered settings, but what I do not understand is why these IDs are not procedurally generated when a server is run.  Is there some need for these to remain persistent between sessions?  It seems much nicer to pass a generator to the ```@setting``` decorator rather than numbering the settings by hand.  Or, for that matter, modify the ```@setting``` decorator to do this itself and remove the ID as an argument altogether.

**Comment by chrismacklin** [comment]
Well, I just answered my own question, but I really feel that this is a bug.

First, run a server with a numbered setting.  In another session,

```
>>> cxn.myserver.mysetting
LabRAD Setting: "myserver" >> "mysetting" (ID=100)
...
```
Quit server, change setting number, restart server.
```python
>>> cxn.refresh()
>>> cxn.myserver.mysetting
...
Error: (4) Server 'myserver' does not have setting 100 [payload=None]
```

**Comment by DanielSank** [comment]
It does seem like that's a bug in the client's `.refresh()`.

**Comment by maffoo** [comment]
That is definitely a bug if refresh isn't working; I thought we had fixed it in #83, but there could still be issues. Are you on master?

As to your real question, it's true that we could change things so that the string setting names are canonical and the numerical IDs are autogenerated, it's just not how we did things. If I can recall back into the misty past when we did this originally, we wanted numerical IDs for servers and settings to save bytes on the wire (instead of sending strings) and we wanted these IDs to be stable so other clients and servers could look them up once and cache them (at least until the server in question restarts, which maybe still needs some work).

**Comment by DanielSank** [comment]
The other thing to keep in mind (maybe) is that if we were to use a network description layer instead of having clients discover settings at run time, static setting ID's would be a requirement.

**Comment by chrismacklin** [comment]
Looks like the issue here goes to the heart of assumptions about settings.  When a setting's ```_refresh()``` method is called, it calls
```info = self._mgr.getSettingInfo(self._server.ID, self.ID)```
making the implicit assumption that a setting's ID never changes.  Seems dangerous; if I were to refactor a server I would need to ensure that settings have static IDs and that all of these IDs have never been used before in any prior version of the server, unless the entire labrad system has been restarted...

**Comment by DanielSank** [comment]
PR's are greatly appreciated :-D

**Comment by chrismacklin** [comment]
I'll take a crack at this, I think I see a fix.

**Comment by chrismacklin** [comment]
Is the entire module ```manager.py``` deprecated?  ```labrad.connect()``` returns a backend.ManagerService object.  What is ```manager.AsyncManager``` there for?

**Comment by maffoo** [comment]
I think if you try to hold on to a particular setting wrapper object, there might be problems; but if you do `cxn.myserver.mysetting` then it should all get hooked up properly (the server wrapper creates new setting wrapper objects after refresh). The unit tests added in #83 are supposed to check exactly the case you describe of a server being modified to renumber settings and then restarted.

**Comment by chrismacklin** [comment]
The unit tests in #83 do not address this, as they re-use the same setting numbers 1 and 2.  I suspect if you change one of them to a different number, the tests will fail.  I can't confirm this as I'm having issues running the labrad tests.  The problem seems to stem from:
```
build/lib/labrad/test/test_client.py:20: in _get_tester
    self.assertTrue(hasattr(self.cxn, 'python_test_server'))
E   AssertionError: False is not true
```

**Comment by ejeffrey** [comment]
Do other clients (Delphi, Java, maybe labview) actually look up setting IDs
automatically, or are they hard-coded in the stub methods?

Evan

**Comment by maffoo** [comment]
Delphi and Java/Scala clients look up settings IDs based on the string names and then cache those IDs. It's been too long since I've touched LabVIEW, but as I recall the VIs for making requests would accept either a string or numerical id for a setting to call and it would look up IDs if you passed a string and do no caching locally (basically incurring one extra round trip to the manager per request).

**Comment by maffoo** [comment]
fixed by 7692b44a
